subsystem/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+
+/// Top-level subsystem manager.
+///
+/// An application is composed of a number of long-lived subsystems. The [Manager] type starts
+/// and manages the life cycle of the subsystems. Whenever a subsystem exits, all other subsystems
+/// are requested to terminate and the manager is shut down.
+pub struct Manager {
+    // Manager name
+    name: &'static str,
+
+    // Order all subsystems to shutdown
+    shutdown_request_tx: broadcast::Sender<()>,
+    shutdown_request_rx: broadcast::Receiver<()>,
+
+    // Signal a subsystem is shutting down. This is taken as an instruction for all subsystems to
+    // shut down. Shutdown completion is detected by all senders having closed this channel.
+    shutting_down_tx: mpsc::Sender<()>,
+    shutting_down_rx: mpsc::Receiver<()>,
+}
+
+impl Manager {
+    /// Initialise a new subsystem manager.
+    pub fn new(name: &'static str) -> Self {
+        log::info!("Initialising subsystem manager {}", name);
+
+        let (shutdown_request_tx, shutdown_request_rx) = broadcast::channel(1);
+        let (shutting_down_tx, shutting_down_rx) = mpsc::channel(1);
+
+        Self {
+            name,
+            shutdown_request_tx,
+            shutdown_request_rx,
+            shutting_down_tx,
+            shutting_down_rx,
+        }
+    }
+
+    /// Build a new subsystem
+    ///
+    /// This gives more control over starting subsystems than [Manager::start] or
+    /// [Manager::start_passive]. Use if extra parameter tuning is required.
+    pub fn builder(&self) -> Builder<'_> {
+        Builder::new(self)
+    }
+
+    /// Start a subsystem. See [Builder::start].
+    pub fn start<T: 'static + Send, F: 'static + Send + Future<Output = ()>>(
+        &self,
+        name: &'static str,
+        subsystem: impl 'static + Send + FnOnce(CallRequest<T>, ShutdownRequest) -> F,
+    ) -> Subsystem<T> {
+        self.builder().with_name(name).start(subsystem)
+    }
+
+    /// Start a passive subsystem. See [Builder::start_passive].
+    pub fn start_passive<T: 'static + Send>(&self, name: &'static str, obj: T) -> Subsystem<T> {
+        self.builder().with_name(name).start_passive(obj)
+    }
+
+    async fn wait_for_subsystems_to_shut_down(mut shutting_down_rx: mpsc::Receiver<()>) {
+        // Wait for the subsystems to go down, signalled by closing the shutting_down channel.
+        while let Some(()) = shutting_down_rx.recv().await {}
+    }
+
+    /// Run the application main task.
+    ///
+    /// Completes when all the subsystems are fully shut down.
+    pub async fn main(mut self) {
+        // The main task just waits for the shutdown signal and coordinates the cleanup at the end.
+        // All other functionality is performed by subsystems.
+
+        // Install the Ctrl-C handler as a subsystem.
+        #[cfg(not(loom))]
+        self.start("ctrl-c", |mut call_rq, mut shutdown_rq| async move {
+            tokio::select! {
+                ctrl_c_signal = tokio::signal::ctrl_c() => {
```

Comment:
What will happen when the user presses ctrl-c? Nothing?

Reply:
Can I test this out? It seems that there is no actual handling of the signal. Where is the code that will cause the subsystem to terminate? As far as I can tell `tokio::signal::ctrl_c` is a future that is resolved when ctrl_c is pressed, but we don't do anything here once that happens (except for log if the future resolved to an error).